sympy/core/function.py

```diff
@@ -261,6 +261,13 @@ def nargs(self):
         # problems with trying to import FiniteSet there
         return FiniteSet(*self._nargs) if self._nargs else S.Naturals0
 
+    def _valid_nargs(self, n : int) -> bool:
+        """ Return True if the specied interger is a valid number of arguments """
+        if self._nargs:
+            return n in self._nargs
+        else:
+            return n in self.nargs
```

Comment:
I think you are right that it always is Naturals0 in the sympy code on github. But someone could have subclassed and modified nargs to return something else. So I implemented this on the safe side.
If we want to change it to 
```
else:
   return True
```
that is fine with me.

Reply:
I feel like it should still be possible to do this in a simpler way that doesn't depend on nargs for the Naturals0 case (at least for most common classes). Bear in mind that performance mainly matters only for the fast path when the number of arguments is valid rather than the error case. Examples of functions having `nargs = Naturals0` are logical functions like `And`, `Or` etc:
```
In [1]: And.nargs
Out[1]: ℕ₀
```

Reply:
`And` is not a `Function` (although it perhaps could be in the future if `Function` were disassociated from `Expr`).

Reply:
> I feel like it should still be possible to do this in a simpler way that doesn't depend on nargs for the Naturals0 case (at least for most common classes). Bear in mind that performance mainly matters only for the fast path when the number of arguments is valid rather than the error case. Examples of functions having `nargs = Naturals0` are logical functions like `And`, `Or` etc:
> 
> ```
> In [1]: And.nargs
> Out[1]: ℕ₀
> ```

Not sure what you mean by simpler. The check is straightforward right? Unless we remove the Naturals0 and/or FiniteSet (which would break compatibility) I do not see how to get this simpler. 

What is perhaps not so nice is that there is both a `_nargs` (a `set`) and `nargs` (a `FiniteSet`), but that is unchanged here.

Reply:
By "simpler" I mean that this bound checking should just be done with plain Python types at least in the common case. The `n in self.nargs` looks simple until you realise how slow it is when `self.nargs` is a symbolic set. Maybe there does need to be a fallback in case someone has defined `nargs` to be something else; I haven't looked through how it all works to understand in detail.

What I do know is that for all cases in the SymPy codebase it should be possible to resolve this without checking `n in self.Naturals0`. There should be some way to avoid doing `n in self.nargs` for common cases like `And` and `Or`.

Reply:
@oscarbenjamin I see what you mean now. The lookup into `Naturals0` is not that bad, although it certainly is not very fast
```
In [14]:  %timeit 3 in Naturals0
1.77 µs ± 5.46 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
Some options to avoid dealing the `Naturals0` altogether (for the common case):

* In `And` and `Or` redefine `def nargs` to something like
```
def _valid_nargs(self, n : int) -> bool:
   return isinstance(n, int) and n>=0 
```
That is fast, but special for `And` and `Or` (and any other method we use this for)
```
n=3
%timeit isinstance(n, int) and n >=0
72.7 ns ± 0.112 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

* Add a similar check inside the regular `nargs`, e.g.
```
def _valid_nargs(self, n : int) -> bool:
        """ Return True if the specied interger is a valid number of arguments """
        if self._nargs:
            return n in self._nargs
        else:
            if isinstance(self.nargs, Naturals0):
                return isinstance(n, int) and n>=0
            else:
                  return n in self.nargs
```
Also fast, but the code does not look very nice.

Either of the options above or the PR as is, is fine with me. And perhaps someone has an even better idea.

Reply:
Since `n` is just `len(args)` there is no need to check if it is an `int` or if it is `>=0`. Those are both guaranteed to be true.

Reply:
> Since `n` is just `len(args)` there is no need to check if it is an `int` or if it is `>=0`. Those are both guaranteed to be true.

You are right. Since `_valid_nargs` is private we can assume a correct `n` is passed, so we could simplify to just `return True`.
More important: is either one of the suggested solutions something you would want in the PR?

Reply:
Maybe something like:
```python
_nargs = self._nargs
if _nargs and n in self._nargs:
    return True
nargs = self.nargs
return nargs is Naturals0 or n in nargs
```
I'm still wondering though if we can just get rid of this altogether and rely on `cls.eval` to just raise if the wrong number of args are given. Is there any potential backwards compatibility problem with that?

Reply:
@oscarbenjamin I updated the code with your suggestion. Skipping the check will cause `cls.eval` to generate a reasonable error message (I checked for a few functions), but in the sympy code there is this comment:
```
        if not cls._valid_nargs(n):
            # XXX: exception message must be in exactly this format to
            # make it work with NumPy's functions like vectorize(). See,
```
If we let `cls.eval` generate the error message, then it will not be in the format required for numpy.

Reply:
That comment about numpy has been sitting there for years. I've wondered whether it is actually accurate or not. I just traced it back through the blame and got to #678 which claims that this is something to do with vectorize. Hopefully numpy has improved in the last 10 years and we can get rid of this now.

Reply:
@oscarbenjamin I created PR #23658, which just removes the check. Most test pass and `np.vectorize` seems to work. So I think we can safely remove the comment.

Both PRs work for me. I have a slight preference for this PR because the error message is a bit more informative and backwards compatibility is 100%.

Reply:
I think if we can skip the check for all classes in the SymPy codebase (which I think this PR does) then that's fine.

Reply:
Experiments with fully removing the check are in #23658. Completely removing the check is not backwards compatible, so the PR is closed in favor of this PR.